### ISSUE TITLE
@typeOf doesn't work on unreachable

### ISSUE THREAD

**Issue by Hejsil** [ziglang/zig#1627]
```rust
test "" {
    // error: unreachable code
    @import("std").debug.assert(@typeOf(unreachable) == noreturn);
}
```

**Comment by thejoshwolfe** [comment]
This might actually be working as intended. The sizeof argument is being evaluated at comptime, and comptime code hit unreachable, right? That's supposed to be a compile error.

**Comment by Hejsil** [comment]
@thejoshwolfe
@andrewrk and I had a back and forth about this before I posted this issue
```
<andrewrk> that's right - @typeOf() takes an expression 
<andrewrk> and evaluating an expression can't hit unreachable at compile-time
<Hejsil> Well, @typeOf doesn't evaluate the expression though 
<Hejsil> And isn't unreachable of type noreturn? 
<andrewrk> it does evaluate the expression 
<Hejsil> How does that work for runtime values? 
<Hejsil> var a: usize = 0; @typeOf(a); works 
<Hejsil> but a can't be comptime evaluated 
<andrewrk> hm I think "evaluation" is an ambiguous term in zig 
<andrewrk> I'm thinking 
<Hejsil> Well, Zig can clearly typecheck unreachable, so it's kinda weird that @typeOf doesn't work 
<andrewrk> hmm I think you're right 
```

TLDR:  `@typeOf` shouldn't need to evaluate the expression to figure out its type. You can `@typeOf` runtime variables, and that works fine, so why shouldn't `unreachable`.

**Comment by thejoshwolfe** [comment]
`@typeOf(if (isWindows()) handle else if (isLinux()) fd else unreachable)`

**Comment by Hejsil** [comment]
Well, shouldn't that work? If we are not windows or linux, then `@typeOf` should return `noreturn` as that is the type of `unreachable`. If you want the code to be `comptime` executed, then store it in a `const` global first.
```
const Handel = if (isWindows()) handle else if (isLinux()) fd else unreachable;
@typeOf(Handle);
```

**Comment by Hejsil** [comment]
Hmmm, actually, I kinda see the problem. We basicly have to special case `unreachable` for `@typeOf` for this to work. Still, it's kinda weird not being able to do `@typeOf(unreachable)`, but I don't have a real usecase other than example code.

```
// This is some documentation:
// As we can see, unreachable is also of type noreturn, which allows it to be part of any expression without effect the resulting type of that expression
@import("std").debug.assert(@typeOf(unreachable) == noreturn);
```

**Comment by thejoshwolfe** [comment]
I'm not able to test this right now, but I would expect a "branch quota exceeded" compile error for `@typeOf(while (true) {})`.

If `@typeOf(expr)` is equivalent to `@typeOf(comptime expr)`, which i think is the case, then it should be impossible to get `noreturn` out of `@typeOf()`.

**Comment by andrewrk** [comment]
`@typeOf(while(true) {} )` hangs the compiler, so there's a bug.

It's correct that `@typeOf(x)` is equivalent to `@typeOf(comptime x)` currently. I'm not sure that's the best behavior though. For example, this asks the question, "why does `@typeOf(variable)` work?"

To which the answer is, another bug:

```zig
test "@typeOf" {
    var a: i32 = undefined;
    @compileLog(comptime a);
}
```

prints

```
| (runtime value)
```

what a mess!

**Comment by Hejsil** [comment]
This probably intended behavior, but the expr passed to `@typeOf` is emitted into the binary. `@typeOf(while(true) {})` doesn't hang the compiler, but runs an inf loop at runtime.
```zig
const std = @import("std");

pub fn main() void {
    var a = true;
    const A = @typeOf(if (a) unreachable else std.os.exit(0));
    std.debug.warn("{}\n", @typeName(A));
}
```
